### PR TITLE
C++: Improve cpp/unsigned-difference-expression-compared-zero

### PR DESCRIPTION
**Pull request by MathiasVP**
Part of https://github.com/github/codeql-c-analysis-team/issues/210.

I've added some tests that demonstrate the false positives we see in the wild on LGTM. One of the false positives can be removed by SimpleRangeAnalysis, and another one can be removed using a `GuardCondition`, so I've added those checks to the query as well. Hopefully, that's not very controversial.

### CODE REVIEW

**Comment by criemen** [comment]
This is a weird test failure:
```
Could not resolve library dependencies in /home/runner/work/semmle-code/semmle-code/ql/cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted: java.lang.NullPointerException
java.lang.NullPointerException
[2036/4203] FAILED(COMPILATION) /home/runner/work/semmle-code/semmle-code/ql/cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/IntegerOverflowTainted.qlref
java.lang.NullPointerException
[2037/4203] FAILED(COMPILATION) /home/runner/work/semmle-code/semmle-code/ql/cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/ArithmeticTainted.qlref
java.lang.NullPointerException
[2038/4203] FAILED(COMPILATION) /home/runner/work/semmle-code/semmle-code/ql/cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/DB-CHECK
java.lang.NullPointerException
[2039/4203] FAILED(COMPILATION) /home/runner/work/semmle-code/semmle-code/ql/cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/CONSISTENCY/unusedLocations.ql
java.lang.NullPointerException
```

**Comment by MathiasVP** [comment]
Agreed. I restarted it. Let's wait and see!

**Comment by MathiasVP** [comment]
Hmm. It seems to be consistently failing

File: cpp/ql/test/experimental/query-tests/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero/test.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+	if(x - y > 0) { } // GOOD [FALSE POSITIVE]
+
+	int n = getAnInt();
+	if (n > x - y) { n = x - y; }
+	if (n > 0) {
+  	y += n; // NOTE: `n` is at most `x - y` at this point.
+  	if (x - y > 0) {} // GOOD [FALSE POSITIVE]
+	}
+}
```
Comment (geoffw0) [inline cpp/ql/test/experimental/query-tests/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero/test.cpp:77]: Fantastic tests!

I doubt we need to pass all of these, but it's good to know where our limits are.

File: cpp/ql/src/experimental/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql [hunk 2]
```diff
@@ -13,11 +13,28 @@
 
 import cpp
 import semmle.code.cpp.commons.Exclusions
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+import semmle.code.cpp.controlflow.Guards
+
+/** Holds if `sub` will never be nonnegative. */
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql:20]:
```suggestion
/** Holds if `sub` will never be negative. */
```

**Review by geoffw0 — COMMENTED** [review]
LGTM aside from a typo.

**Review by geoffw0 — APPROVED** [review]
LGTM.

@criemen do you have anything to add before this gets merged?

**Comment by criemen** [comment]
I've not reviewed the code, I only dealt with the failing tests which were not related to this PR at all. I have no opinion about the PR, if the tests pass and you reviewed it, it sure can go in :+1:

**Pull request merged**